epoxy-adapter/src/main/java/com/airbnb/epoxy/EpoxyVisibilityTracker.java

```diff
@@ -74,47 +92,114 @@ public void detach(@NonNull RecyclerView recyclerView) {
     attachedRecyclerView = null;
   }
 
-  private void processChildren() {
+  private void processChangeEvent(String debug) {
+    processChildren(null, debug);
+  }
+
+  private void processChangeEventWithDetachedView(@Nullable View detachedView, String debug) {
+    processChildren(detachedView, debug);
+  }
+
+  private void processChildren(@Nullable View detachedView, String debug) {
+
+    // On every every events lookup for a new adapter
+    processNewAdapterIfNecessary();
+
+    // Process the detached child if any
+    if (detachedView != null) {
+      processChild(detachedView, true, debug);
+    }
+
+    // Process all attached children
     final RecyclerView recyclerView = attachedRecyclerView;
     if (recyclerView != null) {
       for (int i = 0; i < recyclerView.getChildCount(); i++) {
         final View child = recyclerView.getChildAt(i);
-        if (child != null) {
-          processChild(child);
+        if (child != null && child != detachedView) {
+          // Is some case the detached child is still in the recycler view. Don't process it as it
+          // was already processed.
+          processChild(child, false, debug);
         }
       }
     }
   }
 
-  private void processChild(@NonNull View child) {
-    processChild(child, false);
+  /**
+   * If there is a new adapter on the attached RecyclerView it will resister the data observer and
+   * clear the current visibility states
+   */
+  private void processNewAdapterIfNecessary() {
+    if (attachedRecyclerView != null && attachedRecyclerView.getAdapter() != null) {
+      if (lastAdapterSeen != attachedRecyclerView.getAdapter()) {
+        if (lastAdapterSeen != null) {
+          // Unregister the old adapter
+          lastAdapterSeen.unregisterAdapterDataObserver(this.observer);
+        }
+        // Register the new adapter
+        attachedRecyclerView.getAdapter().registerAdapterDataObserver(this.observer);
+        lastAdapterSeen = attachedRecyclerView.getAdapter();
+        // Clear our visibility items
+        visibilityIdToItemMap.clear();
```

Comment:
it may not be necessary to clear the maps - if `swapAdapter` is used on the recyclerview then the views can be reused with a new adapter.

we can't know if this is the case. the downside to not clearing them is possibly wasted memory (maybe leaks?) and longer lookups, but it might also prevent duplicate visibility events if views are reused and stay on screen

i feel like we should leave it up to the user to tell us that they want to clear view info somehow, since that's the only real way we can know the intention

Reply:
How would you let the user in control of that ? Do you propose explicit `clear()` in the public api ?

I am not a fan of the way it discover a new adapter but I could not find a better way. And I doubt I can detect if the new adapter come from a call to `swapAdapter `

Reply:
Yes, I think an explicit `clear` method could work. It isn't too important though, maybe as a followup? I feel like it's better to give control than to potentially break visibility events for swapped adapters.

I don't think there is a better way to detect a new adapter, so what you're doing seems fine.

Reply:
Ok, make sense. I added `clearVisibilityStates()`.